alibi/explainers/experimental/counterfactuals.py

```diff
@@ -0,0 +1,1264 @@
+
+        self._check_scale(scale)
+
+        if X is not None:
+
+            if self.scale:
+                # infer median absolute deviation (MAD) and update loss
+                if scale == 'median' or isinstance(scale, bool):
+                    scaling_factor = median_abs_deviation(X)
+                    self.backend.distance_fcn.keywords['feature_scale'] = scaling_factor
+
+            if constrain_features:
+                # infer feature ranges and update counterfactual constraints
+                feat_min, feat_max = np.min(X, axis=0), np.max(X, axis=0)
+                self.backend.solution_constraint = [feat_min, feat_max]
+
+        self.fitted = True
+        scaling_method = 'median' if self.scale else 'N/A'
+        self.params = {
+            'scale_loss': self.scale,
+            'scaling_method': scaling_method,
+            'constrain_features': constrain_features,
+            'fitted': self.fitted,
+        }
+
+        return self
+
+    def _check_scale(self, scale: Union[Literal['median'], bool]) -> None:
+        """
+        Checks whether scaling should be performed depending on user input.
+
+        Parameters
+        ----------
+        scale
+            User options for scaling.
+        """
+
+        scale_ = False  # type: bool
+        if isinstance(scale, str):
+            if scale not in WACTHER_CF_VALID_SCALING:
+                logger.warning(f"Received unrecognised option {scale} for scale. No scaling will take place. "
+                               f"Recognised scaling methods are: {WACTHER_CF_VALID_SCALING}.")
+            else:
+                scale_ = True
+
+        if isinstance(scale, bool):
+            if scale:
+                logger.info(f"Defaulting to median absolute deviation scaling!")
+                scale_ = True
+
+        if scale_:
+            loss_params = signature(self.backend.loss_spec['distance']).parameters
+            if 'feature_scale' not in loss_params:
+                logger.warning(
+                    f"Scaling option specified but the loss specified did not have a parameter named 'feature_scale'. "
+                    f"Scaling will not be applied!"
+                )
+                scale_ = False
+
+        self.scale = scale_  # type: bool
+
+
+class WachterCounterfactual(Explainer, FitMixin):
+
+    # TODO: ALEX: TBD: AS WE IMPLEMENT COUNTERFACTUALS, WE MAY REALISE THIS IS COMMON AND WOULD DO THE FOLLOWING:
+    #  - make this a superclass (e.g., CounterfactualAPI). The public classes (e.g., WachterCounterfactual) inherit from
+    #  it to show the docs specific to the method and potentially override some behaviour (eg. fit might have other
+    #  set of arguments, etc)
+    def __init__(self,
+                 predictor: Union[Callable, 'tf.keras.Model', 'keras.Model'],
+                 predictor_type: Literal['blackbox', 'whitebox'] = 'blackbox',
```

Comment:
I don't like the idea of "inferring" because it is usually brittle and does not generalise well. The users of the library should specify what use case they have.